# TeleCo Company: Data Analysis

## Loading Data

In [1]:
import sys
sys.path.insert(0,'../scripts/')

In [2]:
import pandas as pd

In [3]:
from data_loader import load_df_from_csv, load_df_from_excel

In [4]:
fields_description = load_df_from_csv("../data/fields_description.csv")
pd.set_option('display.max_colwidth', None)
fields_description.head()

Fields  \
0  bearer id   
1  Dur. (ms)   
2      Start   
3   Start ms   
4        End   

                                                             Description  
0                                                 xDr session identifier  
1                                      Total Duration of the xDR (in ms)  
2                          Start time of the xDR (first frame timestamp)  
3  Milliseconds offset of start time for the xDR (first frame timestamp)  
4                             End time of the xDR (last frame timestamp)

In [5]:
telData = load_df_from_excel("../data/teleCo_data.xlsx", na_values=['undefined'])
pd.set_option('max_column', None)
telData.head()

Bearer Id               Start  Start ms                 End  \
0  13114483460844900352 2019-04-04 12:01:18     770.0 2019-04-25 14:35:31   
1  13114483482878900224 2019-04-09 13:04:04     235.0 2019-04-25 08:15:48   
2  13114483484080500736 2019-04-09 17:42:11       1.0 2019-04-25 11:58:13   
3  13114483485442799616 2019-04-10 00:31:25     486.0 2019-04-25 07:36:35   
4  13114483499480700928 2019-04-12 20:10:23     565.0 2019-04-25 10:40:32   

   End ms  Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0   662.0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1   606.0  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2   652.0  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3   171.0  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4   954.0  1089009.0  2.082014e+14   3.369979e+10  3.540701e+13   

  Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0   9164566995485190             42.0              5.0   
1            L77566A             65.0              5.0   
2            D42335A              NaN              NaN   
3            T21824A              NaN              NaN   
4            D88865A              NaN              NaN   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  HTTP DL (Bytes)  HTTP UL (Bytes)  \
0                   0.0              NaN              NaN   
1                   0.0              NaN              NaN   
2                   0.0              NaN              NaN   
3                   0.0              NaN              NaN   
4                   0.0              NaN              NaN   

   Activity Duration DL (ms)  Activity Duration UL (ms)   Dur. (ms).1  \
0                    37624.0                    38787.0  1.823653e+09   
1                      168.0                     3560.0  1.365104e+09   
2                        0.0                        0.0  1.361763e+09   
3                     3330.0                    37882.0  1.321510e+09   
4                        0.0                        0.0  1.089009e+09   

  Handset Manufacturer                 Handset Type  \
0      

### Guidance Sub-tasks

#### Top 10 Handsets used by customers

In [6]:
hand_set = telData['Handset Type'].value_counts()[:10]
hand_set

Huawei B528S-23A                19752
Apple iPhone 6S (A1688)          9419
Apple iPhone 6 (A1586)           9023
Apple iPhone 7 (A1778)           6326
Apple iPhone Se (A1723)          5187
Apple iPhone 8 (A1905)           4993
Apple iPhone Xr (A2105)          4568
Samsung Galaxy S8 (Sm-G950F)     4520
Apple iPhone X (A1901)           3813
Samsung Galaxy A5 Sm-A520F       3724
Name: Handset Type, dtype: int64

#### Top 3 Handset Manufacturers

In [7]:
hand_set_manuf = telData['Handset Manufacturer'].value_counts()[:3]
hand_set_manuf


Apple      59565
Samsung    40839
Huawei     34423
Name: Handset Manufacturer, dtype: int64

#### Top 5 Handset per top 3 Handset Manufacturer

In [8]:
new_df = telData.loc[:,['Handset Type','Handset Manufacturer']]
value = new_df.groupby(['Handset Manufacturer', 'Handset Type']).size()
total_list = pd.Series(dtype='object')
for i in hand_set_manuf.index:
    total_list = total_list.append(value[i])
total_list.sort_values(ascending=False)[:5]

Huawei B528S-23A           19752
Apple iPhone 6S (A1688)     9419
Apple iPhone 6 (A1586)      9023
Apple iPhone 7 (A1778)      6326
Apple iPhone Se (A1723)     5187
dtype: int64

#### Interpretation and Recommendation

The phones widely used in the network are manufactured by the Apple, Samsung and Huawei, contributing for 164,827 users. The most used phone is Hwawei B528S-23A covering 19752 of the phones followed by Apple iPhpone 6S, 6, 7 and Se. The Huawei B528S-23A is favoured by users, so we know that selling similar Huawei phones, make promotions on its newer version if it comes and selling the same phone are profitable. We also can infer that iPhones are the next widely used phones and can implement similar marketing or business ideas to them as well. 

## Variable Identification

For better understanding of the columns, exploring the data description [here](https://docs.google.com/spreadsheets/d/1pcNqeUeIph6xAQzlI54KCvi8HM91SUNeeDbdOq3rvbE/edit?usp=sharing) will help

## Data Understanding

In [9]:
from data_information import DataInfo
explorer = DataInfo(telData)

In [10]:
explorer.get_columns()

Columns Listed in the DataFrame are: 


['Bearer Id',
 'Start',
 'Start ms',
 'End',
 'End ms',
 'Dur. (ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Socia

In [11]:
explorer.get_basic_description()

The DataFrame containes 150001 rows and 55 columns.
Current DataFrame Memory Usage:
68672543
Current DataFrame Memory Usage of columns is :
DataFrame Information: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype         
---  ------                                    --------------   -----         
 0   Bearer Id                                 150001 non-null  object        
 1   Start                                     150000 non-null  datetime64[ns]
 2   Start ms                                  150000 non-null  float32       
 3   End                                       150000 non-null  datetime64[ns]
 4   End ms                                    150000 non-null  float32       
 5   Dur. (ms)                                 150000 non-null  float32       
 6   IMSI                                      149431 non-null  float32       
 7   MSISDN/Nu

In [12]:
explorer.get_description()

DataFrame Description: 


Start ms         End ms     Dur. (ms)          IMSI  \
count  150000.000000  150000.000000  1.500000e+05  1.494310e+05   
mean      499.188263     498.800903  1.046086e+05  2.082016e+14   
std       288.611816     288.097656  8.103762e+04  2.148807e+10   
min         0.000000       0.000000  7.142000e+03  2.040471e+14   
25%       250.000000     251.000000  5.744050e+04  2.082014e+14   
50%       499.000000     500.000000  8.639900e+04  2.082015e+14   
75%       749.000000     750.000000  1.324302e+05  2.082018e+14   
max       999.000000     999.000000  1.859336e+06  2.140743e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.489350e+05  1.494290e+05    122172.000000    122189.000000   
mean    4.188282e+10  4.847454e+13       109.795708        17.662884   
std     2.447443e+12  2.241637e+13       619.782715        84.793526   
min     3.360100e+10  4.400152e+11         0.000000         0.000000   
25%     3.365129e+10  3.546071e+13        32.000000         2.000000   
50%     3.366371e+10  3.572201e+13        45.000000         5.000000   
75%     3.368349e+10  8.611971e+13        70.000000        15.000000   
max     8.823971e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
count            150000.000000            150000.000000   
mean              13300.045898              1770.428589   
std               23971.878906              4625.355469   
min                   0.000000                 0.000000   
25%                  43.000000                47.000000   
50%                  63.000000                63.000000   
75%               19710.750000              1120.000000   
max              378160.000000             58613.000000   

       TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
count                 6.185500e+04                 5.335200e+04   
mean                  2.080991e+07                 7.596587e+05   
std                   1.825665e+08                 2.645305e+07   
min                   2.000000e+00                 1.000000e+00   
25%                   3.565150e+04                 4.694750e+03   
50%                   5.687300e+05                 2.094950e+04   
75%                   3.768308e+06                 8.402025e+04   
max                   4.294426e+09                 2.908226e+09   

       DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
count        149247.000000                   149247.000000   
mean             92.844757                        3.069355   
std              13.038032                        6.215232   
min               0.000000                        0.000000   
25%              91.000000                        0.000000   
50%             100.000000                        0.000000   
75%             100.000000                        4.000000   
max             100.000000                       93.000000   

       250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
count                  149247.000000       149247.000000        149209.000000   
mean                        1.717341            1.609654            98.530144   
std                         4.159538            4.828890             4.634285   
min                         0.000000            0.000000             0.000000   
25%                         0.000000            0.000000            99.000000   
50%                         0.000000            0.000000           100.000000   
75%                         1.000000            0.000000           100.000000   
max                       100.000000           94.000000           100.000000   

       10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
count                  149209.000000                   149209.000000   
mean                        0.776749                        0.147987   
std                         3.225177                        1.624523   
min                         0.000000          

In [13]:
explorer.get_column_description('Bearer Id')

count     150001
unique    134709
top             
freq         991
Name: Bearer Id, dtype: object

In [14]:
# Data Inforamtion regarding null-values and type based on Columns
explorer.get_memory_usage()

Current DataFrame Memory Usage of columns is :


Index                                           128
Bearer Id                                   1200008
Start                                       1200008
Start ms                                     600004
End                                         1200008
End ms                                       600004
Dur. (ms)                                    600004
IMSI                                         600004
MSISDN/Number                                600004
IMEI                                         600004
Last Location Name                          1200008
Avg RTT DL (ms)                              600004
Avg RTT UL (ms)                              600004
Avg Bearer TP DL (kbps)                      600004
Avg Bearer TP UL (kbps)                      600004
TCP DL Retrans. Vol (Bytes)                  600004
TCP UL Retrans. Vol (Bytes)                  600004
DL TP < 50 Kbps (%)                          600004
50 Kbps < DL TP < 250 Kbps (%)               600004
250 Kbps < D

### Task 1.1

#### Number of xDR Sessions,Session duration, the total download (DL) and upload (UL) data, the total data volume (in Bytes) during this session for each application Aggregated per User('MSISDN/Number')


In [15]:
pd.set_option('display.float_format', '{:.2f}'.format)

In [16]:
aggs_by_col = {'Bearer Id': 'count', 
               'Dur. (ms)': 'sum',
               'Total UL (Bytes)': 'sum',
               'Total DL (Bytes)': 'sum',
               'Social Media DL (Bytes)': 'sum',
               'Social Media UL (Bytes)': 'sum',
               'Google DL (Bytes)': 'sum',
               'Google UL (Bytes)': 'sum',
               'Email DL (Bytes)': 'sum',
               'Email UL (Bytes)': 'sum',
               'Youtube DL (Bytes)': 'sum',
               'Youtube UL (Bytes)': 'sum',
               'Netflix DL (Bytes)': 'sum',
               'Netflix UL (Bytes)': 'sum',
               'Gaming DL (Bytes)': 'sum',
               'Gaming UL (Bytes)': 'sum',
               'Other DL (Bytes)': 'sum',
               'Other UL (Bytes)': 'sum',
               'Gaming DL (Bytes)': 'sum',
               'Gaming UL (Bytes)': 'sum',
               }

ex1 = telData.groupby('MSISDN/Number').agg(aggs_by_col)
ex1

Bearer Id  Dur. (ms)  Total UL (Bytes)  Total DL (Bytes)  \
MSISDN/Number                                                                  
33601001472.00              2  297950.00       72157568.00      963392640.00   
33601003520.00              1  134969.00       39306820.00      556659648.00   
33601007616.00              3   86982.00      114608056.00     1765123584.00   
33601009664.00              2  253983.00       60009248.00      555208000.00   
33601011712.00              3  214759.00      140257136.00      847126272.00   
...                       ...        ...               ...               ...   
33789980672.00              2  210389.00       70831424.00     1023862016.00   
33789997056.00              2  149798.00       62448672.00     1132500224.00   
3197020930048.00            1  877385.00       37295916.00      194828064.00   
337000045477888.00          1  253030.00       56652840.00      539635008.00   
882397138649088.00          1  869844.00       60456048.00       78697600.00   

                    Social Media DL (Bytes)  Social Media UL (Bytes)  \
MSISDN/Number                                                          
33601001472.00                   4805052.00                 87648.00   
33601003520.00                   3148004.00                 47619.00   
33601007616.00                   3112699.00                 80137.00   
33601009664.00                   3787122.00                  5790.00   
33601011712.00                   4447339.00                104980.00   
...                                     ...                      ...   
33789980672.00                   4200896.00                 49416.00   
33789997056.00                    676534.00                122218.00   
3197020930048.00                  668596.00                 46628.00   
337000045477888.00                496337.00                 25229.00   
882397138649088.00               1500145.00                 45943.00   

                    Google DL (Bytes)  Google UL (Bytes)  Email DL (Bytes)  \
MSISDN/Number                                                                
33601001472.00             7534820.00         2189048.00        3666221.00   
33601003520.00             3343483.00           99643.00        2436500.00   
33601007616.00            19666432.00         8511676.00        4426426.00   
33601009664.00            11993496.00         3269638.00        6513834.00   
33601011712.00            15382479.00         6837538.00        2017004.00   
...                               ...                ...               ...   
33789980672.00             5035730.00         5210743.00        4750550.00   
33789997056.00             8306584.00         4654390.00        2994379.00   
3197020930048.00           8572779.00         1865881.00         842279.00   
337000045477888.00         8215537.00         1615080.00        2989663.00   
882397138649088.00         5985089.00         3233558.00        2518425.00   

                    Email UL (Bytes)  Youtube DL (Bytes)  Youtube UL (Bytes)  \
MSISDN/Number                                                                  
33601001472.00             972922.00         20224452.00         13832319.00   
33601003520.00             768880.00          2137272.00         19196298.00   
33601007616.00            1163713.00         14733094.00         33777228.00   
33601009664.00            1334762.00         12169742.00         22177380.00   
33601011712.00            1557518.00          8090922.00         37233152.00   
...                              ...                 ...                 ...   
33789980672.00             564777.00         20353018.00         17659794.00   
33789997056.00             526633.00         13891172.00         32608242.00   
3197020930048.00           678492.00          9839889.00          2120016.00   
337000045477888.00         328919.00         16690728.00         20044212.00   
882397138649088.00         812549.00         18980320.00         21

## Data Cleaning and Manipulation

In [17]:
from data_cleaner import DataCleaner
cleaner = DataCleaner(explorer.df)

#### Fixing Columns

In [18]:
# changing datatypes
explorer.get_dataframe_columns_unique_value_count()

Unique Value Count
IMSI                                                      44
UL TP > 300 Kbps (%)                                      64
50 Kbps < UL TP < 300 Kbps (%)                            69
250 Kbps < DL TP < 1 Mbps (%)                             74
50 Kbps < DL TP < 250 Kbps (%)                            85
10 Kbps < UL TP < 50 Kbps (%)                             86
DL TP > 1 Mbps (%)                                        86
UL TP < 10 Kbps (%)                                       99
DL TP < 50 Kbps (%)                                      101
Handset Manufacturer                                     170
Avg RTT UL (ms)                                          723
End ms                                                  1001
Start ms                                                1001
Nb of sec with 37500B < Vol UL                          1105
Handset Type                                            1396
Nb of sec with 6250B < Vol UL < 37500B                  1711
Avg RTT DL (ms)                                         2131
Nb of sec with 1250B < Vol UL < 6250B                   3399
IMEI                                                    4583
Nb of sec with 31250B < Vol DL < 125000B                5756
Nb of sec with 6250B < Vol DL < 31250B                  6173
Nb of sec with 125000B < Vol DL                         6459
Avg Bearer TP UL (kbps)                                14529
Nb of sec with Vol DL < 6250B                          22271
Nb of sec with Vol UL < 1250B                          22977
MSISDN/Number                                          35091
TCP UL Retrans. Vol (Bytes)                            35121
Avg Bearer TP DL (kbps)                                41754
Last Location Name                                     45548
TCP DL Retrans. Vol (Bytes)                            54399
Social Media UL (Bytes)                                59078
HTTP UL (Bytes)                                        59943
HTTP DL (Bytes)                                        63524
Dur. (ms)                                              89526
End                                                   100963
Activity Duration DL (ms)                             102561
Activity Duration UL (ms)                             106293
Start                                                 111882
Dur. (ms).1                                           121321
Bearer Id                                             134709
Email UL (Bytes)                                      138699
Social Media DL (Bytes)                               146856
Email DL (Bytes)                                      146916
Google UL (Bytes)                                     147267
Total UL (Bytes)                                      148878
Google DL (Bytes)                                     149024
Youtube UL (Bytes)                                    149279
Other UL (Bytes)                                      149284
Netflix UL (Bytes)                                    149313
Gaming UL (Bytes)                                     149316
Netflix DL (Bytes)                                    149333
Youtube DL (Bytes)                                    149342
Gaming DL (Bytes)                                     149477
Other DL (Bytes)                                      149485
Total DL (Bytes)                                      149497

In [19]:
#removing single valued columns
single_value_cols = explorer.get_dataframe_columns_unique_value_count()
cleaner.remove_single_value_columns(single_value_cols)

#### Standardizing Values

In [20]:
#changing bytes to megabytes
cleaner.convert_bytes_to_megabytes(['Total UL (Bytes)', 'Total DL (Bytes)', 'Social Media DL (Bytes)', 'Social Media UL (Bytes)',
'Google DL (Bytes)', 'Google UL (Bytes)', 'Email DL (Bytes)', 'Email UL (Bytes)', 'Youtube DL (Bytes)', 'Youtube UL (Bytes)', 'Netflix DL (Bytes)', 'Netflix UL (Bytes)', 'Gaming DL (Bytes)','Gaming UL (Bytes)', 'Other DL (Bytes)', 'Other UL (Bytes)', 'Gaming DL (Bytes)', 'Gaming UL (Bytes)', 'Total Lost Packet (Bytes)'])

failed to change values to megabytes


Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18    770.00   
1       13114483482878900224 2019-04-09 13:04:04    235.00   
2       13114483484080500736 2019-04-09 17:42:11      1.00   
3       13114483485442799616 2019-04-10 00:31:25    486.00   
4       13114483499480700928 2019-04-12 20:10:23    565.00   
...                      ...                 ...       ...   
149996   7277825670196679680 2019-04-29 07:28:42    451.00   
149997   7349883264234609664 2019-04-29 07:28:42    483.00   
149998  13114483573367300096 2019-04-29 07:28:43    283.00   
149999  13114483573367300096 2019-04-29 07:28:43    696.00   
150000                                       NaT       NaN   

                       End  End ms  Dur. (ms)               IMSI  \
0      2019-04-25 14:35:31  662.00 1823652.00 208201442131968.00   
1      2019-04-25 08:15:48  606.00 1365104.00 208201911894016.00   
2      2019-04-25 11:58:13  652.00 1361762.00 208200318058496.00   
3      2019-04-25 07:36:35  171.00 1321509.00 208201408577536.00   
4      2019-04-25 10:40:32  954.00 1089009.00 208201408577536.00   
...                    ...     ...        ...                ...   
149996 2019-04-30 06:02:33  214.00   81230.00 208202197106688.00   
149997 2019-04-30 10:41:33  187.00   97970.00 208201911894016.00   
149998 2019-04-30 10:46:12  810.00   98249.00 208201710567424.00   
149999 2019-04-30 10:40:34  327.00   97910.00 208202096443392.00   
150000                 NaT     NaN        NaN                NaN   

        MSISDN/Number              IMEI Last Location Name  Avg RTT DL (ms)  \
0      33664962560.00 35521208254464.00   9164566995485190            42.00   
1      33681854464.00 35794009980928.00            L77566A            65.00   
2      33760626688.00 35281512169472.00            D42335A              NaN   
3      33750343680.00 35356611182592.00            T21824A              NaN   
4      33699794944.00 35407009939456.00            D88865A              NaN   
...               ...               ...                ...              ...   
149996 33650688000.00 35483111391232.00            D20434A            32.00   
149997 33663449088.00 35660509478912.00            D10223C            27.00   
149998 33621891072.00 35721209446400.00            T51102A            43.00   
149999 33619621888.00 86186202759168.00            L88342B            37.00   
150000            NaN               NaN                NaN              NaN   

        Avg RTT UL (ms)  Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                  5.00                    23.00                    44.00   
1                  5.00                    16.00                    26.00   
2                   NaN                     6.00                     9.00   
3                   NaN                    44.00                    44.00   
4                   NaN                     6.00                     9.00   
...                 ...                      ...                      ...   
149996             0.00                    52.00                    65.00   
149997             2.00                    23.00                    54.00   
149998             6.00                    43.00                    47.00   
149999             5.00                    34.00                    37.00   
150000              NaN                      NaN                      NaN   

        TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                               NaN                          NaN   
1                               NaN                          NaN   
2                               NaN                          NaN   
3                               NaN                          NaN   
4                               NaN                          NaN   
...                             ...                          ...   
149996                          NaN                          NaN   
149997                          NaN    

#### Filtering Data

In [21]:
cleaner.remove_duplicates()

#### Fixing Missing Values

In [22]:
explorer = DataInfo(cleaner.df)

In [23]:
explorer.get_columns_with_missing_values()

['Start',
 'Start ms',
 'End',
 'End ms',
 'Dur. (ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Total UL (MegaBytes

In [24]:
explorer.get_total_missing_values()

The total number of missing values is 1048375
12.71 % missing values.


1048375

In [25]:
explorer.get_column_based_missing_values()

missing_count            type
Start                                                 1  datetime64[ns]
Start ms                                              1         float32
End                                                   1  datetime64[ns]
End ms                                                1         float32
Dur. (ms)                                             1         float32
IMSI                                                570         float32
MSISDN/Number                                      1066         float32
IMEI                                                572         float32
Last Location Name                                 1153          object
Avg RTT DL (ms)                                   27829         float32
Avg RTT UL (ms)                                   27812         float32
Avg Bearer TP DL (kbps)                               1         float32
Avg Bearer TP UL (kbps)                               1         float32
TCP DL Retrans. Vol (Bytes)                       88146         float32
TCP UL Retrans. Vol (Bytes)                       96649         float32
DL TP < 50 Kbps (%)                                 754         float32
50 Kbps < DL TP < 250 Kbps (%)                      754         float32
250 Kbps < DL TP < 1 Mbps (%)                       754         float32
DL TP > 1 Mbps (%)                                  754         float32
UL TP < 10 Kbps (%)                                 792         float32
10 Kbps < UL TP < 50 Kbps (%)                       792         float32
50 Kbps < UL TP < 300 Kbps (%)                      792         float32
UL TP > 300 Kbps (%)                                792         float32
HTTP DL (Bytes)                                   81474         float32
HTTP UL (Bytes)                                   81810         float32
Activity Duration DL (ms)                             1         float32
Activity Duration UL (ms)                             1         float32
Dur. (ms).1                                           1         float32
Handset Manufacturer                               9559          object
Handset Type                                       9559          object
Nb of sec with 125000B < Vol DL                   97538         float32
Nb of sec with 1250B < Vol UL < 6250B             92894         float32
Nb of sec with 31250B < Vol DL < 125000B          93586         float32
Nb of sec with 37500B < Vol UL                   130254         float32
Nb of sec with 6250B < Vol DL < 31250B            88317         float32
Nb of sec with 6250B < Vol UL < 37500B           111843         float32
Nb of sec with Vol DL < 6250B                       755         float32
Nb of sec with Vol UL < 1250B                       793         float32
Total UL (MegaBytes)                                  1         float64
Total DL (MegaBytes)                                  1         float64

In [26]:
explorer.get_column_based_missing_percentage()

total_missing_values  \
Bearer Id                                                    0   
Start                                                        1   
Start ms                                                     1   
End                                                          1   
End ms                                                       1   
Dur. (ms)                                                    1   
IMSI                                                       570   
MSISDN/Number                                             1066   
IMEI                                                       572   
Last Location Name                                        1153   
Avg RTT DL (ms)                                          27829   
Avg RTT UL (ms)                                          27812   
Avg Bearer TP DL (kbps)                                      1   
Avg Bearer TP UL (kbps)                                      1   
TCP DL Retrans. Vol (Bytes)                              88146   
TCP UL Retrans. Vol (Bytes)                              96649   
DL TP < 50 Kbps (%)                                        754   
50 Kbps < DL TP < 250 Kbps (%)                             754   
250 Kbps < DL TP < 1 Mbps (%)                              754   
DL TP > 1 Mbps (%)                                         754   
UL TP < 10 Kbps (%)                                        792   
10 Kbps < UL TP < 50 Kbps (%)                              792   
50 Kbps < UL TP < 300 Kbps (%)                             792   
UL TP > 300 Kbps (%)                                       792   
HTTP DL (Bytes)                                          81474   
HTTP UL (Bytes)                                          81810   
Activity Duration DL (ms)                                    1   
Activity Duration UL (ms)                                    1   
Dur. (ms).1                                                  1   
Handset Manufacturer                                      9559   
Handset Type                                              9559   
Nb of sec with 125000B < Vol DL                          97538   
Nb of sec with 1250B < Vol UL < 6250B                    92894   
Nb of sec with 31250B < Vol DL < 125000B                 93586   
Nb of sec with 37500B < Vol UL                          130254   
Nb of sec with 6250B < Vol DL < 31250B                   88317   
Nb of sec with 6250B < Vol UL < 37500B                  111843   
Nb of sec with Vol DL < 6250B                              755   
Nb of sec with Vol UL < 1250B                              793   
Social Media DL (MegaBytes)                                  0   
Social Media UL (MegaBytes)                                  0   
Google DL (MegaBytes)                                        0   
Google UL (MegaBytes)                                        0   
Email DL (MegaBytes)                                         0   
Email UL (MegaBytes)                                         0   
Youtube DL (MegaBytes)                                       0   
Youtube UL (MegaBytes)                                       0   
Netflix DL (MegaBytes)                                       0   
Netflix UL (MegaBytes)                                       0   
Gaming DL (MegaBytes)                                        0   
Gaming UL (MegaBytes)                                        0   
Other DL (MegaBytes)                                         0   
Other UL (MegaBytes)                                         0   
Total UL (MegaBytes)                                         1   
Total DL (MegaBytes)                                         1   

                                         missing_percentage  
Bearer Id                                             0.0 %  
Start                                                 0.0 %  
Start ms                                              0.0 %  
End                                                   0.0 %  
End ms                             

In [27]:
explorer.get_columns_missing_percentage_greater_than(30.0)

Index(['TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)',
       'HTTP DL (Bytes)', 'HTTP UL (Bytes)', 'Nb of sec with 125000B < Vol DL',
       'Nb of sec with 1250B < Vol UL < 6250B',
       'Nb of sec with 31250B < Vol DL < 125000B',
       'Nb of sec with 37500B < Vol UL',
       'Nb of sec with 6250B < Vol DL < 31250B',
       'Nb of sec with 6250B < Vol UL < 37500B'],
      dtype='object')

In [28]:
cleaner.df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype         
---  ------                                    --------------   -----         
 0   Bearer Id                                 150001 non-null  object        
 1   Start                                     150000 non-null  datetime64[ns]
 2   Start ms                                  150000 non-null  float32       
 3   End                                       150000 non-null  datetime64[ns]
 4   End ms                                    150000 non-null  float32       
 5   Dur. (ms)                                 150000 non-null  float32       
 6   IMSI                                      149431 non-null  float32       
 7   MSISDN/Number                             148935 non-null  float32       
 8   IMEI                                      149429 non-null  float32       
 9   Last Location N

In [29]:
cleaner.remove_unwanted_columns(explorer.get_columns_missing_percentage_greater_than(30.0))
cleaner.df.shape

(150001, 45)

In [30]:
cleaner.df.drop([cleaner.df.index[-1]], inplace=True)
cleaner.df.shape

(150000, 45)

In [31]:
explorer = DataInfo(cleaner.df)

In [32]:
explorer.get_columns_with_missing_values()

['IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B']

In [33]:
explorer.get_column_based_missing_values()

missing_count     type
IMSI                                      569  float32
MSISDN/Number                            1065  float32
IMEI                                      571  float32
Last Location Name                       1152   object
Avg RTT DL (ms)                         27828  float32
Avg RTT UL (ms)                         27811  float32
DL TP < 50 Kbps (%)                       753  float32
50 Kbps < DL TP < 250 Kbps (%)            753  float32
250 Kbps < DL TP < 1 Mbps (%)             753  float32
DL TP > 1 Mbps (%)                        753  float32
UL TP < 10 Kbps (%)                       791  float32
10 Kbps < UL TP < 50 Kbps (%)             791  float32
50 Kbps < UL TP < 300 Kbps (%)            791  float32
UL TP > 300 Kbps (%)                      791  float32
Handset Manufacturer                     9558   object
Handset Type                             9558   object
Nb of sec with Vol DL < 6250B             754  float32
Nb of sec with Vol UL < 1250B             792  float32

Check if skew for numeric missing values and if not fill missing values with mean value but if skew use median

In [34]:
cleaner.fill_numeric_values(explorer.get_columns_with_missing_values())
cleaner.df.sample(10)

Bearer Id               Start  Start ms  \
121256   7349883261257159680 2019-04-28 04:11:45    587.00   
123573  13114483569153400832 2019-04-28 05:37:07     82.00   
20926   13042425959523600384 2019-04-24 20:15:25    700.00   
145417  13114483573168400384 2019-04-29 06:09:14    555.00   
60430   13042425969387700224 2019-04-26 03:32:28    319.00   
144532  13114483573104199680 2019-04-29 05:36:12    284.00   
104200   7349883258439189504 2019-04-27 08:36:53    774.00   
28202   13114483550247999488 2019-04-24 07:10:54    963.00   
65741    7349883254648160256 2019-04-26 07:30:46     47.00   
7760    12682137987811399680 2019-04-24 14:19:49    996.00   

                       End  End ms  Dur. (ms)               IMSI  \
121256 2019-04-29 08:02:23  432.00  100237.00 208201408577536.00   
123573 2019-04-29 07:24:31  253.00   92844.00 208200703934464.00   
20926  2019-04-25 00:42:23  374.00   16017.00 208201911894016.00   
145417 2019-04-30 08:30:19  781.00   94865.00 208201442131968.00   
60430  2019-04-27 20:46:17  839.00  148429.00 208201408577536.00   
144532 2019-04-30 07:57:01  689.00   94849.00 208201710567424.00   
104200 2019-04-29 09:44:09  503.00  176835.00 208201777676288.00   
28202  2019-04-26 06:49:35  132.00  171520.00 208201710567424.00   
65741  2019-04-27 05:31:07  956.00   79221.00 208202096443392.00   
7760   2019-04-25 00:17:15  436.00   35845.00 208201005924352.00   

        MSISDN/Number              IMEI Last Location Name  Avg RTT DL (ms)  \
121256 33679415296.00 35346809094144.00            T78636A            45.00   
123573 33622841344.00 35482809401344.00            D74543A           160.00   
20926  33629345792.00 35281411506176.00            L52579C            48.00   
145417 33659621376.00 35332909170688.00            D92711A            75.00   
60430  33699182592.00 86376900984832.00            D18587B            66.00   
144532 33782212608.00 35530909679616.00            D40009A            23.00   
104200 33698287616.00 35215808397312.00            L61032A            45.00   
28202  33668812800.00 35307110006784.00            L40084B            24.00   
65741  33761050624.00 35832106844160.00            D14136B            27.00   
7760   33762275328.00 86032607346688.00            G50907B           492.00   

        Avg RTT UL (ms)  Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
121256             5.00                    43.00                    43.00   
123573             3.00                    48.00                    48.00   
20926             10.00                 65009.00                  1841.00   
145417             0.00                    43.00                    46.00   
60430             29.00                 43544.00                  1250.00   
144532             5.00                    46.00                    50.00   
104200             5.00                    48.00                    53.00   
28202              5.00                    44.00                    51.00   
65741              2.00                    61.00                    51.00   
7760               8.00                   257.00                  1150.00   

        DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
121256               100.00                            0.00   
123573               100.00                            0.00   
20926                 85.00                            3.00   
145417               100.00                            0.00   
60430                 82.00                            3.00   
144532               100.00                            0.00   
104200               100.00                            0.00   
28202                100.00                            0.00   
65741                100.00                            0.00   
7760                 100.00                            0.00   

        250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  \
121256                           0.00                0.00   
123573                           0.00                

In [35]:
explorer = DataInfo(cleaner.df)
explorer.get_columns_with_missing_values()

['Last Location Name', 'Handset Manufacturer', 'Handset Type']

In [36]:
cleaner.fill_non_numeric_values(explorer.get_columns_with_missing_values(),ffill=True,bfill=True)

C:\Python38\lib\site-packages\pandas\core\frame.py:4462: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18    770.00   
1       13114483482878900224 2019-04-09 13:04:04    235.00   
2       13114483484080500736 2019-04-09 17:42:11      1.00   
3       13114483485442799616 2019-04-10 00:31:25    486.00   
4       13114483499480700928 2019-04-12 20:10:23    565.00   
...                      ...                 ...       ...   
149995  13042425979329300480 2019-04-29 07:28:41    615.00   
149996   7277825670196679680 2019-04-29 07:28:42    451.00   
149997   7349883264234609664 2019-04-29 07:28:42    483.00   
149998  13114483573367300096 2019-04-29 07:28:43    283.00   
149999  13114483573367300096 2019-04-29 07:28:43    696.00   

                       End  End ms  Dur. (ms)               IMSI  \
0      2019-04-25 14:35:31  662.00 1823652.00 208201442131968.00   
1      2019-04-25 08:15:48  606.00 1365104.00 208201911894016.00   
2      2019-04-25 11:58:13  652.00 1361762.00 208200318058496.00   
3      2019-04-25 07:36:35  171.00 1321509.00 208201408577536.00   
4      2019-04-25 10:40:32  954.00 1089009.00 208201408577536.00   
...                    ...     ...        ...                ...   
149995 2019-04-30 00:01:49  407.00   59587.00 208201408577536.00   
149996 2019-04-30 06:02:33  214.00   81230.00 208202197106688.00   
149997 2019-04-30 10:41:33  187.00   97970.00 208201911894016.00   
149998 2019-04-30 10:46:12  810.00   98249.00 208201710567424.00   
149999 2019-04-30 10:40:34  327.00   97910.00 208202096443392.00   

        MSISDN/Number              IMEI Last Location Name  Avg RTT DL (ms)  \
0      33664962560.00 35521208254464.00   9164566995485190            42.00   
1      33681854464.00 35794009980928.00            L77566A            65.00   
2      33760626688.00 35281512169472.00            D42335A            45.00   
3      33750343680.00 35356611182592.00            T21824A            45.00   
4      33699794944.00 35407009939456.00            D88865A            45.00   
...               ...               ...                ...              ...   
149995 33668648960.00 35332510711808.00            T85721A           313.00   
149996 33650688000.00 35483111391232.00            D20434A            32.00   
149997 33663449088.00 35660509478912.00            D10223C            27.00   
149998 33621891072.00 35721209446400.00            T51102A            43.00   
149999 33619621888.00 86186202759168.00            L88342B            37.00   

        Avg RTT UL (ms)  Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                  5.00                    23.00                    44.00   
1                  5.00                    16.00                    26.00   
2                  5.00                     6.00                     9.00   
3                  5.00                    44.00                    44.00   
4                  5.00                     6.00                     9.00   
...                 ...                      ...                      ...   
149995             9.00                 63420.00                  1393.00   
149996             0.00                    52.00                    65.00   
149997             2.00                    23.00                    54.00   
149998             6.00                    43.00                    47.00   
149999             5.00                    34.00                    37.00   

        DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                    100.00                            0.00   
1                    100.00                            0.00   
2                    100.00                            0.00   
3                    100.00                            0.00   
4                    100.00                            0.00   
...                     ...                             ...   
149995                54.00                           23.00   
149996               100.00                            0.00   
149997          

#### Adding Merged and Spliting Columns and Standardizing

In [37]:
#spliting date-time value to separate date and time columns for each start and end. (start_date, start_time, end_date, end_time)
cleaner.separate_date_time_column('Start', 'start')
cleaner.separate_date_time_column('End', 'end')

Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18    770.00   
1       13114483482878900224 2019-04-09 13:04:04    235.00   
2       13114483484080500736 2019-04-09 17:42:11      1.00   
3       13114483485442799616 2019-04-10 00:31:25    486.00   
4       13114483499480700928 2019-04-12 20:10:23    565.00   
...                      ...                 ...       ...   
149995  13042425979329300480 2019-04-29 07:28:41    615.00   
149996   7277825670196679680 2019-04-29 07:28:42    451.00   
149997   7349883264234609664 2019-04-29 07:28:42    483.00   
149998  13114483573367300096 2019-04-29 07:28:43    283.00   
149999  13114483573367300096 2019-04-29 07:28:43    696.00   

                       End  End ms  Dur. (ms)               IMSI  \
0      2019-04-25 14:35:31  662.00 1823652.00 208201442131968.00   
1      2019-04-25 08:15:48  606.00 1365104.00 208201911894016.00   
2      2019-04-25 11:58:13  652.00 1361762.00 208200318058496.00   
3      2019-04-25 07:36:35  171.00 1321509.00 208201408577536.00   
4      2019-04-25 10:40:32  954.00 1089009.00 208201408577536.00   
...                    ...     ...        ...                ...   
149995 2019-04-30 00:01:49  407.00   59587.00 208201408577536.00   
149996 2019-04-30 06:02:33  214.00   81230.00 208202197106688.00   
149997 2019-04-30 10:41:33  187.00   97970.00 208201911894016.00   
149998 2019-04-30 10:46:12  810.00   98249.00 208201710567424.00   
149999 2019-04-30 10:40:34  327.00   97910.00 208202096443392.00   

        MSISDN/Number              IMEI Last Location Name  Avg RTT DL (ms)  \
0      33664962560.00 35521208254464.00   9164566995485190            42.00   
1      33681854464.00 35794009980928.00            L77566A            65.00   
2      33760626688.00 35281512169472.00            D42335A            45.00   
3      33750343680.00 35356611182592.00            T21824A            45.00   
4      33699794944.00 35407009939456.00            D88865A            45.00   
...               ...               ...                ...              ...   
149995 33668648960.00 35332510711808.00            T85721A           313.00   
149996 33650688000.00 35483111391232.00            D20434A            32.00   
149997 33663449088.00 35660509478912.00            D10223C            27.00   
149998 33621891072.00 35721209446400.00            T51102A            43.00   
149999 33619621888.00 86186202759168.00            L88342B            37.00   

        Avg RTT UL (ms)  Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                  5.00                    23.00                    44.00   
1                  5.00                    16.00                    26.00   
2                  5.00                     6.00                     9.00   
3                  5.00                    44.00                    44.00   
4                  5.00                     6.00                     9.00   
...                 ...                      ...                      ...   
149995             9.00                 63420.00                  1393.00   
149996             0.00                    52.00                    65.00   
149997             2.00                    23.00                    54.00   
149998             6.00                    43.00                    47.00   
149999             5.00                    34.00                    37.00   

        DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                    100.00                            0.00   
1                    100.00                            0.00   
2                    100.00                            0.00   
3                    100.00                            0.00   
4                    100.00                            0.00   
...                     ...                             ...   
149995                54.00                           23.00   
149996               100.00                            0.00   
149997          

In [38]:
#merging columns to generate new values
#function for merging the columns, basic subtraction
def merge_sub(col1:pd.Series, col2:pd.Series):
    return col1.sub(col2)

def merge_add(col1:pd.Series, col2:pd.Series):
    return col1.add(col2)
#Total lost packet
cleaner.create_new_columns_from('Total Data (MegaBytes)','Total UL (MegaBytes)','Total DL (MegaBytes)',merge_add)
#Jitter
cleaner.create_new_columns_from('Jitter (ms)','Activity Duration UL (ms)','Activity Duration DL (ms)',merge_sub)
#Avg Delay
cleaner.create_new_columns_from('Avg Delay (ms)','Avg RTT UL (ms)','Avg RTT DL (ms)',merge_sub)
#Avg Throughput
cleaner.create_new_columns_from('Avg Throughput (kbps)','Avg Bearer TP UL (kbps)','Avg Bearer TP DL (kbps)',merge_sub)

Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18    770.00   
1       13114483482878900224 2019-04-09 13:04:04    235.00   
2       13114483484080500736 2019-04-09 17:42:11      1.00   
3       13114483485442799616 2019-04-10 00:31:25    486.00   
4       13114483499480700928 2019-04-12 20:10:23    565.00   
...                      ...                 ...       ...   
149995  13042425979329300480 2019-04-29 07:28:41    615.00   
149996   7277825670196679680 2019-04-29 07:28:42    451.00   
149997   7349883264234609664 2019-04-29 07:28:42    483.00   
149998  13114483573367300096 2019-04-29 07:28:43    283.00   
149999  13114483573367300096 2019-04-29 07:28:43    696.00   

                       End  End ms  Dur. (ms)               IMSI  \
0      2019-04-25 14:35:31  662.00 1823652.00 208201442131968.00   
1      2019-04-25 08:15:48  606.00 1365104.00 208201911894016.00   
2      2019-04-25 11:58:13  652.00 1361762.00 208200318058496.00   
3      2019-04-25 07:36:35  171.00 1321509.00 208201408577536.00   
4      2019-04-25 10:40:32  954.00 1089009.00 208201408577536.00   
...                    ...     ...        ...                ...   
149995 2019-04-30 00:01:49  407.00   59587.00 208201408577536.00   
149996 2019-04-30 06:02:33  214.00   81230.00 208202197106688.00   
149997 2019-04-30 10:41:33  187.00   97970.00 208201911894016.00   
149998 2019-04-30 10:46:12  810.00   98249.00 208201710567424.00   
149999 2019-04-30 10:40:34  327.00   97910.00 208202096443392.00   

        MSISDN/Number              IMEI Last Location Name  Avg RTT DL (ms)  \
0      33664962560.00 35521208254464.00   9164566995485190            42.00   
1      33681854464.00 35794009980928.00            L77566A            65.00   
2      33760626688.00 35281512169472.00            D42335A            45.00   
3      33750343680.00 35356611182592.00            T21824A            45.00   
4      33699794944.00 35407009939456.00            D88865A            45.00   
...               ...               ...                ...              ...   
149995 33668648960.00 35332510711808.00            T85721A           313.00   
149996 33650688000.00 35483111391232.00            D20434A            32.00   
149997 33663449088.00 35660509478912.00            D10223C            27.00   
149998 33621891072.00 35721209446400.00            T51102A            43.00   
149999 33619621888.00 86186202759168.00            L88342B            37.00   

        Avg RTT UL (ms)  Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                  5.00                    23.00                    44.00   
1                  5.00                    16.00                    26.00   
2                  5.00                     6.00                     9.00   
3                  5.00                    44.00                    44.00   
4                  5.00                     6.00                     9.00   
...                 ...                      ...                      ...   
149995             9.00                 63420.00                  1393.00   
149996             0.00                    52.00                    65.00   
149997             2.00                    23.00                    54.00   
149998             6.00                    43.00                    47.00   
149999             5.00                    34.00                    37.00   

        DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                    100.00                            0.00   
1                    100.00                            0.00   
2                    100.00                            0.00   
3                    100.00                            0.00   
4                    100.00                            0.00   
...                     ...                             ...   
149995                54.00                           23.00   
149996               100.00                            0.00   
149997          

In [39]:
cleaner.change_columns_type_to(['start_date','end_date','IMSI','Handset Manufacturer','Handset Type','IMEI','MSISDN/Number','Bearer Id'],'category')

Bearer Id               Start  Start ms  \
0       13114483460844900352 2019-04-04 12:01:18    770.00   
1       13114483482878900224 2019-04-09 13:04:04    235.00   
2       13114483484080500736 2019-04-09 17:42:11      1.00   
3       13114483485442799616 2019-04-10 00:31:25    486.00   
4       13114483499480700928 2019-04-12 20:10:23    565.00   
...                      ...                 ...       ...   
149995  13042425979329300480 2019-04-29 07:28:41    615.00   
149996   7277825670196679680 2019-04-29 07:28:42    451.00   
149997   7349883264234609664 2019-04-29 07:28:42    483.00   
149998  13114483573367300096 2019-04-29 07:28:43    283.00   
149999  13114483573367300096 2019-04-29 07:28:43    696.00   

                       End  End ms  Dur. (ms)               IMSI  \
0      2019-04-25 14:35:31  662.00 1823652.00 208201442131968.00   
1      2019-04-25 08:15:48  606.00 1365104.00 208201911894016.00   
2      2019-04-25 11:58:13  652.00 1361762.00 208200318058496.00   
3      2019-04-25 07:36:35  171.00 1321509.00 208201408577536.00   
4      2019-04-25 10:40:32  954.00 1089009.00 208201408577536.00   
...                    ...     ...        ...                ...   
149995 2019-04-30 00:01:49  407.00   59587.00 208201408577536.00   
149996 2019-04-30 06:02:33  214.00   81230.00 208202197106688.00   
149997 2019-04-30 10:41:33  187.00   97970.00 208201911894016.00   
149998 2019-04-30 10:46:12  810.00   98249.00 208201710567424.00   
149999 2019-04-30 10:40:34  327.00   97910.00 208202096443392.00   

        MSISDN/Number              IMEI Last Location Name  Avg RTT DL (ms)  \
0      33664962560.00 35521208254464.00   9164566995485190            42.00   
1      33681854464.00 35794009980928.00            L77566A            65.00   
2      33760626688.00 35281512169472.00            D42335A            45.00   
3      33750343680.00 35356611182592.00            T21824A            45.00   
4      33699794944.00 35407009939456.00            D88865A            45.00   
...               ...               ...                ...              ...   
149995 33668648960.00 35332510711808.00            T85721A           313.00   
149996 33650688000.00 35483111391232.00            D20434A            32.00   
149997 33663449088.00 35660509478912.00            D10223C            27.00   
149998 33621891072.00 35721209446400.00            T51102A            43.00   
149999 33619621888.00 86186202759168.00            L88342B            37.00   

        Avg RTT UL (ms)  Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                  5.00                    23.00                    44.00   
1                  5.00                    16.00                    26.00   
2                  5.00                     6.00                     9.00   
3                  5.00                    44.00                    44.00   
4                  5.00                     6.00                     9.00   
...                 ...                      ...                      ...   
149995             9.00                 63420.00                  1393.00   
149996             0.00                    52.00                    65.00   
149997             2.00                    23.00                    54.00   
149998             6.00                    43.00                    47.00   
149999             5.00                    34.00                    37.00   

        DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                    100.00                            0.00   
1                    100.00                            0.00   
2                    100.00                            0.00   
3                    100.00                            0.00   
4                    100.00                            0.00   
...                     ...                             ...   
149995                54.00                           23.00   
149996               100.00                            0.00   
149997          

### Data Exploration and Analysis

In [40]:
explorer = DataInfo(cleaner.df)
explorer.get_columns()

Columns Listed in the DataFrame are: 


['Bearer Id',
 'Start',
 'Start ms',
 'End',
 'End ms',
 'Dur. (ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Social Media DL (MegaBytes)',
 'Social Media UL (MegaBytes)',
 'Google DL (MegaBytes)',
 'Google UL (MegaBytes)',
 'Email DL (MegaBytes)',
 'Email UL (MegaBytes)',
 'Youtube DL (MegaBytes)',
 'Youtube UL (MegaBytes)',
 'Netflix DL (MegaBytes)',
 'Netflix UL (MegaBytes)',
 'Gaming DL (MegaBytes)',
 'Gaming UL (MegaBytes)',
 'Other DL (MegaBytes)',
 'Ot